[4-6단계 - 방탈출 사용자 예약] 띠용(주용은) 미션 제출합니다.

jye326 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
시간에 쫒겨 객체 지향을 신경 쓸 여유가 없었다. 리팩토링하며 적용해봐야겠다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [x] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

- 처음 배우는 개념이 많아 시간에 쫒기며 최소 구현 기능 위주로 얼렁뚱땅 구현한 것 같습니다. 학습 예제에 등장한 코드를 계속 재탕하며 코딩한 느낌이라 더 세련되게 개선할 수 있는 방법을 배우고 싶습니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

solarsalt committed:
docs: add PR template

jye326 committed:
test(service) : 예약 생성 테스트 추가

jye326 committed:
refactor(package) : user 관련 패키지 생성

jye326 committed:
feat(controller) : Login 페이지 컨트롤러

jye326 committed:
feat(controller) : Login 페이지 컨트롤러 생성

jye326 committed:
feat(domain) : User 도메인 추가

jye326 committed:
feat(controller) : Login 기능 추가

- 로그인 기능 구현
 - 로그인 컨트롤러
 - 로그인 서비스
 - 로그인 리포지토리 인터페이스 및 Jdbc 구현체

jye326 committed:
feat(dto) : Login 요청 및 응답 추가

jye326 committed:
feat(infra) : Token 응답 및 토큰 처리기 구현

- gradle 의존성 추가
- secret key, 만료 기한 추가

jye326 committed:
feat(resource) : js, html 로그인 업데이트

jye326 committed:
test(controller) : LoginControllerTest

- 사용자 로그인 테스트

jye326 committed:
feat(docs) : README.md 업데이트

- 4단계 요구사항 구현

jye326 committed:
fix(controller) : 예약 가능 시간 조회 API 수정

- themeId 입력 방식 변경

jye326 committed:
test(database) : 테스트용 더미 데이터 추가

jye326 committed:
fix(resource) : 예약 페이지 클라이언트 코드 변경

jye326 committed:
fix(dto) : 예약 요청 형식 변경

jye326 committed:
feat(controller) : 사용자 인증 자동화 구현

- 쿠키 정보 활용 인증
- 예약 요청 파라미터에 자동 주입

jye326 committed:
fix(resource) : 예약 정보 파라미터 수정

jye326 committed:
fix(model) : User 필드 클래스명 변경

jye326 committed:
feat(dto) : 예약 요청 분리

- 관리자, 사용자

jye326 committed:
fix(controller) : 컨트롤러 범위 확대

- Login -> Member 컨트롤러 (로그인 및 회원 관련 기능 담당)

jye326 committed:
feat(controller) : 관리자 컨트롤러 생성

jye326 committed:
feat(controller) : 예약 기능 분리

- 관리자용, 사용자용

jye326 committed:
fix(model) : User -> Member 로 변경

jye326 committed:
refactor(service) : MemberService 포맷팅

jye326 committed:
fix(model) : Role.of 수정

jye326 committed:
fix(app) : user 네이밍 삭제

- ex) userEmail -> email

jye326 committed:
fix(controller) : 예약 추가 함수 rename

jye326 committed:
feat(config, controller) : 관리자 권한 확인 인터셉터 추가

jye326 committed:
refactor(controller) : 관리자 페이지 컨트롤러 requestMapping

jye326 committed:
docs(README) : 리드미 업데이트

jye326 committed:
feat(resources) : 필터링 js 코드 업데이트

jye326 committed:
docs(README) : 리드미 업데이트

jye326 committed:
step2 pulling

jye326 committed:
fix(resources) : sql 스키마 변경

- 예약 스키마 name 필드 -> member_id로 변경

jye326 committed:
feat(controller) : filter 컨트롤러 생성

jye326 committed:
fix(model) : 예약 객체 필드값 변경

- Name -> Member

jye326 committed:
feat(controller) : 관리자용 예약 필터링 기능 구현

jye326 committed:
feat(controller) : 멤버 추가 기능 구현

jye326 committed:
feat(controller) : 멤버 추가 기능 구현

jye326 committed:
test(test) : 전체 테스트 코드 수정

jye326 committed:
pulling

jye326 committed:
delete(dto) : 미사용 클래스 삭제

- LoginResponse.java

jye326 commented on src/main/java/roomescape/dto/LoginResponse.java:
```diff
@@ -0,0 +1,7 @@
+import jakarta.validation.constraints.NotNull;
+
+public record LoginResponse(@Email String email, @NotNull String password) {
+}
```
사용되지 않는 클래스라 삭제했어

jye326 commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -52,6 +52,20 @@ private static void Test_Theme_Post() {
+                .when().post("/themes")
+                .then().log().all()
+                .statusCode(201);
+    }
```
다시확인해볼래?

jye326 committed:
fix(controller) : 관리자 예약 추가 함수 반환 타입 변경

Member 객체 -> Reservation 객체

jye326 commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,46 @@
+        this.memberService = memberService;
+    }
+
+    @PostMapping("/admin/reservations")
+    public ResponseEntity<Member> addReservation(@RequestBody AdminReservationRequest request) {
+
+        Member member = memberService.getMemberById(request.userId());
+        reservationService.addReservation(member,
+                UserReservationRequest.of(request.date(), request.timeId(), request.themeId()));
+        return ResponseEntity.status(HttpStatus.CREATED).body(member);
+    }
```
생성된 예약객체를 반환하도록 수정했어

jye326 committed:
fix(schema) : email 유니크 설정

jye326 commented on src/main/resources/schema.sql:
```diff
@@ -14,33 +14,67 @@ CREATE TABLE theme
+    password VARCHAR(255) NOT NULL,
+    role     VARCHAR(255) NOT NULL,
+    PRIMARY KEY (id)
+);
```
반영했어!

ddaaac commented on src/main/java/roomescape/config/WebMvcConfiguration.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addInterceptor(checkAdminInterceptor)
+                .addPathPatterns("/admin/**");
+    }
+
+    @Override
+    public void addArgumentResolvers(List<HandlerMethodArgumentResolver> argumentResolvers) {
+        argumentResolvers.add(authenticationPrincipalArgumentResolver);
```
Interceptor와 ArgumentResolver는 어떤 차이를 두고 따로 구현해주신 걸까요?

jye326 replied:
인터셉터는 요청을 컨트롤러에 넘기기 이전에 가로채 공통적으로 처리하는 로직을 수행합니다. 현재는 /admin으로 시작되는 요청에 대해 관리자 권한을 확인합니다.
ArgumentResolver는 웹 요청을 자바 코드에서 다루기 위한 객체 파라미터로 변환하여 주입하는 로직을 수행합니다. 현재는 예약 추가 시에 동작하며 요청의 쿠키를 통해 사용자 (Member) 객체를 예약 추가 컨트롤러에 파라미터로 주입합니다.

ddaaac commented on src/main/java/roomescape/controller/AdminController.java:
```diff
@@ -0,0 +1,46 @@
+import roomescape.service.ReservationService;
+
+@RestController
+public class AdminController {
```
adminController를 따로 두는 것과 각 도메인에 /admin을 두는 건 어떤 차이가 있을까요?

jye326 replied:
현재 컨트롤러들을 처리 API 요청 기준으로 분리하고 있어 관리자 요청 API (/admin/**) 를 공통적으로 관리하는 객체가 있어야겠다고 판단했습니다.

ddaaac commented on src/main/java/roomescape/controller/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -0,0 +1,38 @@
+
+import jakarta.servlet.http.Cookie;
+import jakarta.servlet.http.HttpServletRequest;
+import org.apache.coyote.BadRequestException;
```
확인해주세요.

jye326 replied:
확실히 Tomcat의 오류 클래스 보다는 Spring 예외를 사용하는 것이 좋겠네요!
반영했습니다!

ddaaac commented on src/main/java/roomescape/controller/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -0,0 +1,38 @@
+        HttpServletRequest request = webRequest.getNativeRequest(HttpServletRequest.class);
+        Cookie[] cookies = request.getCookies();
+        if (cookies != null) {
+            String userEmail = memberService.extractEmailFromCookies(cookies);
+            return memberService.findByEmail(userEmail);
```
두 단계에 나눠서 할 필요가 있을까 싶어요. cookies -> email을 뽑아내는 게 memberservice의 역할로 적당한지도 생각해보시면 좋겠네요.

jye326 replied:
ArgumentResolver 내부 로직이 여러 단계로 나뉘어 복잡도가 올라간 것 같습니다. 관련 로직을 service에 위임하여 한번에 처리하였습니다. 또한 member service를 auth service와 구분하여 객체 간 역할 분리를 개선했습니다.

ddaaac commented on src/main/java/roomescape/controller/CheckAdminInterceptor.java:
```diff
@@ -0,0 +1,34 @@
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return false;
+        }
+        String email = memberService.extractEmailFromCookies(cookies);
+        Role role = memberService.getRoleByEmail(email);
```
여기서도 뭔가 ArgumentResolver와 중복되는 코드가 있어보여요. 책임이 적당히 나눠져있는지 확인해보면 좋겠네요.

jye326 replied:
ArgumentResolver 동일하게 리팩토링 하였습니다.

ddaaac commented on src/main/java/roomescape/controller/CheckAdminInterceptor.java:
```diff
@@ -0,0 +1,34 @@
+        Cookie[] cookies = request.getCookies();
+        if (cookies == null) {
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return false;
```
예외를 던져서 처리하는 것과 false를 리턴해서 처리하는 방식의 차이는 무엇일까요?

jye326 replied:
요청의 처리 시점 및 위치에 차이가 있는 것 같습니다.

인터셉터가 false를 리턴하면 요청 처리 자체가 중단되고, response.setStatus에서 응답상태를 변경하여 즉시 리턴합니다. 즉 서버코드에 도달하기 이전에 반환됩니다. 예외를 던지면 해당 예외를 처리하기 위해 서버 코드의 예외 핸들러가 동작합니다.

ddaaac commented on src/main/java/roomescape/controller/LoginController.java:
```diff
@@ -0,0 +1,52 @@
+            return ResponseEntity.ok(new NameResponse(name.getValue()));
+        }
+        throw new BadRequestException("인증 실패");
+    }
```
이 코멘트도 확인해주세요.

jye326 replied:
수정하였습니다!

ddaaac commented on src/main/java/roomescape/controller/GlobalExceptionHandler.java:
```diff
@@ -16,9 +17,16 @@ public ResponseEntity<String> handleIllegalArgumentException(IllegalArgumentExce
+                .body("[ERROR] " + ex);
     }
+
+//    @ExceptionHandler(Exception.class)
```
여기는 왜 주석처리한걸까요?

jye326 replied:
깜빡했습니다.. 죄송합니다 😰

ddaaac commented on src/main/java/roomescape/service/MemberService.java:
```diff
@@ -0,0 +1,74 @@
+
+    public Member login(String email, String password) {
+        Member member = memberRepository.login(email, password);
+        System.out.println(member);
```
사용하지 않는 print는 지워주세요.

jye326 replied:
삭제했습니다!

ddaaac commented on src/main/java/roomescape/service/MemberService.java:
```diff
@@ -0,0 +1,74 @@
+        return member;
+    }
+
+    public TokenResponse createToken(String payload) {
+        String accessToken = jwtTokenProcessor.createToken(payload);
+        return new TokenResponse(accessToken);
+    }
+
+    public Cookie createCookie(String token) {
```
쿠키를 굽거나 토큰을 만드는 책임과 MemberService의 책임에 대해 생각해보셔도 좋을 것 같아요.

jye326 replied:
MemberService와 무관하다고 판단하여 관련 기능을 AuthService로 분리하였습니다.

ddaaac commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -49,4 +53,24 @@ public void deleteReservation(long id) {
     }
 
+    public List<Reservation> getFilteredReservation(Long themeId, Long memberId, LocalDate dateFrom, LocalDate dateTo) {
+        List<Reservation> allReservations = reservationRepository.getAllReservations();
+        List<Reservation> filteredReservation = new ArrayList<>(allReservations);
+        if (themeId != null) {
+            filteredReservation = allReservations.stream().filter(r -> r.getTheme().getId().equals(themeId)).toList();
+        }
+        if (memberId != null) {
+            filteredReservation = filteredReservation.stream().filter(r -> r.getMember().getId().equals(memberId))
+                    .toList();
+        }
+        if (dateFrom != null) {
+            filteredReservation = filteredReservation.stream()
+                    .filter(r -> !r.getReservationDateTime().getDate().isBefore(dateFrom)).toList();
+        }
+        if (dateTo != null) {
+            filteredReservation = filteredReservation.stream()
+                    .filter(r -> !r.getReservationDateTime().getDate().isAfter(dateTo)).toList();
+        }
```
reservation이 많아질수록 findAll과 같은 풀스캔의 성능은 많이 나빠집니다. 가능하면 쿼리로 풀어보면 좋을 것 같아요.

jye326 replied:
동적 쿼리로 수정했습니다!

ddaaac reviewed (changes requested):
안녕하세요 띠용 👋 

구현 잘 해주셨어요.

몇 가지 코멘트 남겼으니 코드래빗의 코멘트들과 함께 확인해주세요.

jye326 committed:
refactor(package) : 패키지 구조 변경

jye326 committed:
fix(service) : 인증 관련 기능 분리 및 인증 토큰 키 값 수정

- MemberService -> AuthService
- 토큰 키 값 token -> loginToken

jye326 committed:
feat(controller): AdminController 매핑 및 경로 리팩터링

 - AdminController의 공통 경로를 @RequestMapping으로 리팩터링
 - POST 및 GET 매핑 경로 간소화
 - WebMvcConfiguration 내 생성자 파라미터 순서 변경
 - 불필요한 주석 및 출력문 제거

jye326 committed:
fix(auth): 쿠키를 통한 회원 조회 로직 단순화

- `AuthenticationPrincipalArgumentResolver`에서 `authService.getMemberFromCookies()`로 로직 변경
- `AuthService`에 `getMemberByEmail` 및 `getMemberFromCookies` 메서드 추가
- 로직 캡슐화를 위해 `getPayload` 접근 제한자를 private으로 변경

jye326 committed:
refactor(infra): CheckAdminInterceptor 간소화

- AuthService의 getRoleFromCookies 메서드 추가 및 활용
- 중복된 메서드 제거 및 코드 구조 개선

jye326 committed:
fix(controller):인증 실패 시 예외 처리 수정

- 로그 추가로 인증 실패 원인 기록
- BadRequestException 제거 및 UNAUTHORIZED 응답 반환

jye326 committed:
feat(repository, service): 예약 필터 기능 추가

 - ReservationRepository 메서드에 필터링 조건 추가 (themeId, memberId, dateFrom, dateTo).
 - JdbcReservationRepository에서 필터링 조건에 따라 동적 SQL 및 파라미터 로직 추가.
 - MemoryReservationRepository에 새로운 메서드 시그니처 적용.
 - 서비스 계층 메서드에 필터링 로직 연동.
 - ReservationServiceTest 수정으로 새로운 getReservations 메서드 테스트 반영.
 - 필요 없는 MemberService 메서드 제거로 코드 정리.

jye326 reviewed (commented):
다음 단계 미션을 진행하느라 리뷰 반영이 늦었습니다! 죄송합니다!!

jye326 committed:
fix(auth): MemberService 의존성 제거, exception handling 예외 수정

 - MemberService 의존성을 제거하여 코드 간결화
 - BadRequestException을 HttpClientErrorException으로 변경하고 적절한 상태코드 추가

ddaaac commented on src/main/java/roomescape/infra/auth/CheckAdminInterceptor.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+        Cookie[] cookies = request.getCookies();
+        if (cookies == null) {
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return false;
+        }
+        if (authService.getRoleFromCookies(cookies) != Role.ADMIN) {
+            response.setStatus(HttpServletResponse.SC_FORBIDDEN);
```
응답의 상태코드를 정하고, 메세지를 추가하는 레이어는 `@Controller`와 `@ControllerAdvice`로 한정하는 건 어떨까요?
여기서 예외를 던지면 일관되게 처리할 수 있을 것 같아요.

ddaaac reviewed (approved):
안녕하세요 띠용 👋 

리뷰 반영 잘 해주셨네요.

다음 단게도 화이팅입니다!
